README.md
[![CI workflow](https://img.shields.io/github/workflow/status/dns3l/sra/main?label=ci&logo=github)](https://github.com/dns3l/sra/actions/workflows/main.yml)
[![GitHub release](https://img.shields.io/github/release/dns3l/sra.svg&logo=github)](https://github.com/dns3l/sra/releases/latest)
[![Semantic Release](https://img.shields.io/badge/semantic--release-angular-e10079?logo=semantic-release)](https://github.com/semantic-release/semantic-release)
![License](https://img.shields.io/github/license/dns3l/sra)

## [Smallstep][1] registration authority for DNS3L

`docker pull ghcr.io/dns3l/sra`

[1]: https://smallstep.com/docs/registration-authorities/acme-for-certificate-manager

### Configuration

| variable | note | default |
| --- | --- | --- |
| ENVIRONMENT | `production` or other deployments | |
| SRA_BIND | Registration Authority Bind Port or Address | `:9443` |
| SRA_DNS | Registration Authority DNS Names | `"localhost", "acmera"` |
| STEP_CA_URL | Certificate Manager Authority URL | `https://stepca:9000` |
| STEP_CA_FINGERPRINT | Certificate Manager Authority Fingerprint | `foobar` |
| STEP_CA_PROVISIONER | Certificate Manager JWK Provisioner Name | `acme-ra` |
| STEP_CA_PASSWORD | JWK provisioner password | random |
| SRA_DATABASE | MariaDB database name | `acmera` |
| SRA_DB_USER | database user | `acmera` |
| SRA_DB_PASS | user password | random |
| SRA_DB_HOST | MariaDB server IP/FQDN | `db` |
| SRA_RESOLVER | Optional DNS resolver IP (1.2.3.4:53) | |
| MARIADB_ROOT_PASSWORD | MariaDB root password | |

If `ENVIRONMENT` is `! production` and `MARIADB_ROOT_PASSWORD` is set the database and user are created.

Mount a custom step-ca config to `/etc/stepca.conf.json` if environment based template seems not sufficient.


CHANGELOG.md
## [1.0.3](https://github.com/dns3l/sra/compare/v1.0.2...v1.0.3) (2023-02-22)


### Bug Fixes

* Configurable DSN resolver ([#21](https://github.com/dns3l/sra/issues/21)) ([49f66b3](https://github.com/dns3l/sra/commit/49f66b3fc0019dbc3d5bb25f65f2783d09417b44))

## [1.0.2](https://github.com/dns3l/sra/compare/v1.0.1...v1.0.2) (2022-12-06)


### Bug Fixes

* workaround issue [#7](https://github.com/dns3l/sra/issues/7) ([#13](https://github.com/dns3l/sra/issues/13)) ([811a4dc](https://github.com/dns3l/sra/commit/811a4dc9092b22accddf739b51a9314329e9e222))

## [1.0.1](https://github.com/dns3l/sra/compare/v1.0.0...v1.0.1) (2022-11-17)


### Bug Fixes

* Add dockerize ([#5](https://github.com/dns3l/sra/issues/5)) ([2ee3bb4](https://github.com/dns3l/sra/commit/2ee3bb45b97cdc23fad835ba65b01efbf52a230f))

# 1.0.0 (2022-08-23)


### Features

* Try initial release ([aa5022c](https://github.com/dns3l/sra/commit/aa5022cd7100b936737629bc516a03a659ef47fe))


docker-entrypoint.sh
#!/bin/bash
set -e

umask 0022

# usage: file_env VAR [DEFAULT]
#    ie: file_env 'XYZ_DB_PASSWORD' 'example'
#  (will allow for "$XYZ_DB_PASSWORD_FILE" to fill in the value of
#   "$XYZ_DB_PASSWORD" from a file, especially for Docker's secrets feature)
function file_env() {
  local var="$1"
  local fileVar="${var}_FILE"
  local def="${2:-}"
  if [ "${!var:-}" ] && [ "${!fileVar:-}" ]; then
    echo >&2 "error: both $var and $fileVar are set (but are exclusive)"
    exit 1
  fi
  local val="$def"
  if [ "${!var:-}" ]; then
    val="${!var}"
  elif [ "${!fileVar:-}" ]; then
    val="$(< "${!fileVar}")"
  fi
  export "$var"="$val"
  unset "$fileVar"
}

# envs=(
#   XYZ_API_TOKEN
# )
# haveConfig=
# for e in "${envs[@]}"; do
#   file_env "$e"
#   if [ -z "$haveConfig" ] && [ -n "${!e}" ]; then
#     haveConfig=1
#   fi
# done

# return true if specified directory is empty
function directory_empty() {
  [ -n "$(find "${1}"/ -prune -empty)" ]
}

function random_token() {
  tr -cd '[:alnum:]' </dev/urandom | fold -w32 | head -n1
}

SERVICE_TIMEOUT=${SERVICE_TIMEOUT:-300s} # wait for dependencies

echo Running: "$@"

export SRA_BIND=${SRA_BIND:-":9443"}
export SRA_DNS=${SRA_DNS:-'"localhost","sra","acmera"'}
export SRA_RESOLVER=${SRA_RESOLVER:-""}

export STEP_CA_URL=${STEP_CA_URL:-"https://stepca:9000"}
export STEP_CA_FINGERPRINT=${STEP_CA_FINGERPRINT:-"foobar"}
export STEP_CA_PROVISIONER=${STEP_CA_PROVISIONER:-"acme-ra"}
export STEP_CA_PASSWORD=${STEP_CA_PASSWORD:-$(random_token)}

export SRA_DATABASE=${SRA_DATABASE:-"sra"}
export SRA_DB_USER=${SRA_DB_USER:-"sra"}
export SRA_DB_PASS=${SRA_DB_PASS:-$(random_token)}
export SRA_DB_HOST=${SRA_DB_HOST:-"db"}

production=false
if [[ ${ENVIRONMENT,,} == "production" ]]; then
  production=true
fi

# Avoid destroying bootstrapping by simple start/stop
if [[ ! -e ${STEPPATH}/.bootstrapped ]]; then
  ### list none idempotent code blocks, here...

  touch ${STEPPATH}/.bootstrapped
fi

###
### ACME RA DB bootstrapping...
###

if [[ "${production}" == "false" && -n "${MARIADB_ROOT_PASSWORD}" ]]; then
  echo "Bootstrapping ACME RA Database..."
  set +e
  /dckrz -wait tcp://${SRA_DB_HOST}:3306 -timeout ${SERVICE_TIMEOUT} -- \
    echo "quit" | mariadb -uroot -p"${MARIADB_ROOT_PASSWORD}" -h"${SRA_DB_HOST}" -D"${SRA_DATABASE}"
  if [ "$?" != "0" ]; then # create DB
    set -e
    echo "Create ${SRA_DATABASE}..."
    /dckrz -wait tcp://${SRA_DB_HOST}:3306 -timeout ${SERVICE_TIMEOUT} -- \
      cat <<EOSQL | mariadb -uroot -p"${MARIADB_ROOT_PASSWORD}" -h"${SRA_DB_HOST}"
CREATE DATABASE IF NOT EXISTS ${SRA_DATABASE};
CREATE USER IF NOT EXISTS ${SRA_DB_USER}@'%' IDENTIFIED BY '${SRA_DB_PASS}';
GRANT ALL ON ${SRA_DATABASE}.* TO ${SRA_DB_USER}@'%';
FLUSH PRIVILEGES;
EOSQL
  else # change password (optionally)
    set -e
    echo "Change password ${SRA_DB_PASS}..."
    /dckrz -wait tcp://${SRA_DB_HOST}:3306 -timeout ${SERVICE_TIMEOUT} -- \
      cat <<EOSQL | mariadb -uroot -p"${MARIADB_ROOT_PASSWORD}" -h"${SRA_DB_HOST}"
ALTER USER IF EXISTS ${SRA_DB_USER}@'%' IDENTIFIED BY '${SRA_DB_PASS}';
FLUSH PRIVILEGES;
EOSQL
  fi
fi

###
### ACME RA bootstrapping...
###

echo -n "${STEP_CA_PASSWORD}" >${STEPPATH}/.acme-ra.pass

echo "Creating ACME RA configuration..."
mkdir -p -m 0700 ${STEPPATH}/config
mkdir -p -m 0700 ${STEPPATH}/certs

# mkdir -p -m 0700 ${STEPPATH}/db
# "db": {
#   "type": "badgerV2",
#   "dataSource": "${STEPPATH}/db" },

if [ -r /etc/stepca.conf.json -a -s /etc/stepca.conf.json ]; then
  ln -fs /etc/stepca.conf.json ${STEPPATH}/config/ca.json
else
  /dckrz -template ${STEPPATH}/ca.json.tmpl:${STEPPATH}/config/ca.json
fi

# Workaround for https://github.com/dns3l/sra/issues/7
# /dckrz -wait ${STEP_CA_URL} -skip-tls-verify -wait-http-status-code 401 -timeout ${SERVICE_TIMEOUT} -- \
step ca bootstrap -f --ca-url ${STEP_CA_URL} --fingerprint ${STEP_CA_FINGERPRINT} # --install

/dckrz -wait tcp://${SRA_DB_HOST}:3306 -timeout ${SERVICE_TIMEOUT} -- echo "Ok. MariaDB is there."

if [[ `basename ${1}` == "step-ca" ]]; then # prod
  if [ -z "${SRA_RESOLVER}" ]; then
    exec "$@" </dev/null #>/dev/null 2>&1
  else
    exec "$@" --resolver ${SRA_RESOLVER} </dev/null #>/dev/null 2>&1
  fi
else # dev
  step-ca ${STEPPATH}/config/ca.json --issuer-password-file ${STEPPATH}/.acme-ra.pass || true
fi

# fallthrough...
exec "$@"
